### PR TITLE
Replication

### PR DESCRIPTION
**Pull request by alainjobart**
@enisoc 
OK I've been fighting with this for a bit. Unit test passes now.
But I feel we have two issues:
- we don't have any unit test that locks a keyspace in thread A, tries to lock a keyspace in thread B, unlocks the keyspace in thread A, and make sure thread B now has the lock. I have a feeling that wouldn't work for etcd (need this for keyspace, shard, and srvshard). The one that wouldn't work is srvshard.
- and that feeling is because of this line in lock.go line 168:
				if resp.Action == "delete" {
which I think should be:
				if resp.Action == "comparAndDelete" {

I'll work on adding the unit test, we'll see.

### CODE REVIEW

File: go/vt/etcdtopo/client_test.go [hunk 1]
```diff
@@ -185,8 +254,49 @@ func (c *fakeClient) SetCluster(machines []string) bool {
...
+	c.Unlock()
+
+	n.mu.Lock()
+	index := n.watchIndex
+	n.watchIndex++
+	n.watches[index] = forwarder
+	n.mu.Unlock()
+
+	// and wait until we stop, each action will write to forwarder, send
+	// these along. The order in select is critical: we only want to
```
Comment (enisoc) [inline go/vt/etcdtopo/client_test.go]: If the order matters, you need to use two selects. Each select chooses pseudo-randomly with uniform distribution if more than one case is ready to proceed.

File: go/vt/etcdtopo/serving_graph.go [hunk 2]
```diff
@@ -231,6 +239,75 @@ func (s *Server) UpdateTabletEndpoint(cell, keyspace, shard string, tabletType t
...
+	go func() {
+		// get the current version of the file
+		ep, modifiedVersion, err := s.getEndPoints(cellName, keyspace, shard, tabletType)
+		if err != nil {
+			// node doesn't exist
+			modifiedVersion = 0
+			ep = nil
+		}
+
+		// re-check for stop here to be safe, in case the
```
Comment (enisoc) [inline go/vt/etcdtopo/serving_graph.go]: Since select is pseudorandom, this re-check for stop is only 50% effective.

**Comment by enisoc** [comment]
LGTM

**Comment by alainjobart** [comment]
@enisoc 
Please look at the last commit, will do the same for SrvShard (in the not-should-exist case).

**Comment by alainjobart** [comment]
OK added tests for Shard and SrvShard, and found and fixed the expected bug on delete vs compareAndDelete. So ready to merge after LGTM.

**Comment by enisoc** [comment]
LGTM

**Pull request merged**